src/s3/async_put_object.rs

```diff
@@ -0,0 +1,232 @@
+    // the pub_object operation.
+    Closing(Pin<Box<dyn Future<Output = Result<PutObjectOutput, SdkError<PutObjectError>>> + 'a>>),
+    // We have completed writing to s3.
+    Closed,
+}
+
+/// Implements the `AsyncWrite` trait and writes data to S3 using the `put_object` API.
+///
+/// # Example
+///
+/// ```no_run
+/// use aws_config;
+/// use cobalt_aws::s3::{get_client, AsyncPutObject};
+/// use futures::AsyncWriteExt;
+///
+/// # tokio_test::block_on(async {
+/// let shared_config = aws_config::load_from_env().await;
+/// let client = get_client(&shared_config).unwrap();
+///
+/// let mut writer = AsyncPutObject::new(&client, "my-bucket", "my-key");
+/// writer.write_all(b"File contents").await.unwrap();
+///
+/// // The contents are pushed to S3 when the `.close()` method is called.
+/// writer.close().await.unwrap();
+/// # })
+/// ```
+pub struct AsyncPutObject<'a> {
+    client: &'a Client,
+    key: String,
+    bucket: String,
+    buf: Vec<u8>,
+    state: PutObjectState<'a>,
+}
+
+impl<'a> AsyncPutObject<'a> {
+    /// Create a new `AsyncPutObject` which will write data to the given `bucket` and `key`.
+    pub fn new(client: &'a Client, bucket: &str, key: &str) -> Self {
+        AsyncPutObject {
+            client,
+            key: key.into(),
+            bucket: bucket.into(),
+            buf: vec![],
+            state: PutObjectState::Writing,
+        }
+    }
+}
+
+impl<'a> AsyncWrite for AsyncPutObject<'a> {
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<Result<usize, Error>> {
+        match self.state {
+            PutObjectState::Writing => {
+                self.buf.extend(buf);
+                Poll::Ready(Ok(buf.len()))
+            }
+            _ => Poll::Ready(Err(Error::new(
+                ErrorKind::Other,
+                "Attempted to .write() writer after .close().",
+            ))),
+        }
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Result<(), Error>> {
+        match self.state {
+            PutObjectState::Writing => Poll::Ready(Ok(())),
+            _ => Poll::Ready(Err(Error::new(
+                ErrorKind::Other,
+                "Attempted to .flush() writer after .close().",
+            ))),
+        }
+    }
+
+    fn poll_close<'b>(
+        mut self: Pin<&'b mut AsyncPutObject<'a>>,
+        cx: &'b mut Context<'_>,
+    ) -> Poll<Result<(), Error>> {
+        match self.state {
+            PutObjectState::Writing => {
+                // Create the put_object future and transition to
+                // the Closing state.
+                let fut = self
+                    .client
+                    .put_object()
+                    .bucket(&self.bucket)
+                    .key(&self.key)
+                    .body(mem::take(&mut self.buf).into())
+                    .acl(ObjectCannedAcl::BucketOwnerFullControl)
```

Comment:
Would it be reasonable for callers to want to customize the ACL here? (Maybe for the future though).

Reply:
Yeah, definitely. I was pondering if it makes sense for us to let the user build their own request object and then we just populate the `.body` 🤔

Let's see how it goes in practice and iterate for sure 👍